### ISSUE TITLE
 Incorrect LoRA Layers being Fine-Tuned in Code vs. Paper

### ISSUE THREAD

**Issue by euminds** [snap-research/weights2weights#5]
In the paper, it is mentioned that the key and value projection matrices are fine-tuned using LoRA
![image](https://github.com/user-attachments/assets/12682168-7dbc-4e2c-8a1f-d88206420584)

However, the code is currently fine-tuning the query (to_q) and value (to_v) matrices instead, which is not aligned with the method described in the paper.

weight_dimensions = torch.load("files/weight_dimensions.pt")

Results:
'base_model.model.down_blocks.0.attentions.0.transformer_blocks.0.attn1.to_q.lora_A.weight':
[torch.Size([320]), torch.Size([1, 320])]
'base_model.model.down_blocks.0.attentions.0.transformer_blocks.0.attn1.to_q.lora_B.weight':
[torch.Size([320]), torch.Size([320, 1])]
'base_model.model.down_blocks.0.attentions.0.transformer_blocks.0.attn1.to_v.lora_A.weight':
[torch.Size([320]), torch.Size([1, 320])]
'base_model.model.down_blocks.0.attentions.0.transformer_blocks.0.attn1.to_v.lora_B.weight':
[torch.Size([320]), torch.Size([320, 1])]
......

**Comment by avdravid** [comment]
Hi, thanks for pointing that out. It appears to be a typo. I'll make sure to correct that. We follow the PEFT implementation: https://github.com/huggingface/peft/blob/main/examples/lora_dreambooth/train_dreambooth.py

**Comment by euminds** [comment]
In addition, we observe that the size of the tensor "V" is [99648, 10000], while the tensor produced by combining all LoRA of ID images has a size of [99648, 64974]. I have a few questions regarding this:
Is "V" obtained by performing PCA on the [99648, 64974] tensor?
If so, how is PCA implemented in the code? For example, which specific libraries or functions are used?
If "V" is not obtained through PCA on the [99648, 64974] tensor, what is the process by which it is generated?
Thank you for your assistance!

**Comment by euminds** [comment]
> In addition, we observe that the size of the tensor "V" is [99648, 10000], while the tensor produced by combining all LoRA of ID images has a size of [99648, 64974]. I have a few questions regarding this:
> Is "V" obtained by performing PCA on the [99648, 64974] tensor?
> If so, how is PCA implemented in the code? For example, which specific libraries or functions are used?
> If "V" is not obtained through PCA on the [99648, 64974] tensor, what is the process by which it is generated?
> Thank you for your assistance!

In addition, we observe that the size of the tensor "V" is [99648, 10000], while the tensor produced by combining all LoRA of ID images has a size of [99648, 64974]. I have a few questions regarding this:
Is "V" obtained by performing PCA on the [99648, 64974] tensor?
If so, how is PCA implemented in the code? For example, which specific libraries or functions are used?
If "V" is not obtained through PCA on the [99648, 64974] tensor, what is the process by which it is generated?
Thank you for your assistance!

**Comment by avdravid** [comment]
> In addition, we observe that the size of the tensor "V" is [99648, 10000], while the tensor produced by combining all LoRA of ID images has a size of [99648, 64974]. I have a few questions regarding this: Is "V" obtained by performing PCA on the [99648, 64974] tensor? If so, how is PCA implemented in the code? For example, which specific libraries or functions are used? If "V" is not obtained through PCA on the [99648, 64974] tensor, what is the process by which it is generated? Thank you for your assistance!


You are correct. The "V" tensor corresponds to the v matrix when doing SVD/PCA. You can run the following below: 

a = torch.load("data.pt")
mean = torch.mean(a, dim=0)
std = torch.std(a, dim=0)
scaled_data = (a - mean)/std
u,s,v = torch.svd(scaled_data)


If you get out-of-memory issues when running SVD on large tensors, you can also use torch.svd_lowrank.
